### PR TITLE
Draft: Adds services capabilities

### PR DESCRIPTION
**Pull request by Markcial**
Related: https://github.com/nektos/act/issues/173 

Hi, this is a draft MR. Will need some more work

The network is created and removed by name, causing collisions, and it is created only once, I am not sure if that is the right setup, maybe it might be needed to be different on each job.

Containers and services should be included on that network, it's done here but maybe there is a better way. This MR also lacks proper testing, some advice on this would be good.

The configuration on the services are done in a hacky way, they should have properly merged environment settings, and some other configuration values.

I can help on any further changes or if you feel that someone else can't take better care of it feel free to modify it.

Thanks for this project

### CODE REVIEW

File: pkg/runner/run_context.go [hunk 1]
```diff
@@ -150,6 +183,21 @@ func (rc *RunContext) startJobContainer() common.Executor {
...
+		return container.NewDockerNetworkCreateExecutor(defaultNetwork)(ctx)
+	}
+}
+
+func (rc *RunContext) removeNetwork() common.Executor {
+	return func(ctx context.Context) error {
+		return container.NewDockerNetworkRemoveExecutor(defaultNetwork)(ctx)
+	}
+}
+
```
Comment (catthehacker) [inline pkg/runner/run_context.go:200]: If another workflow references same services, there will be a clash.
Perhaps network should be named `act-<workflow-name>-network`

File: pkg/runner/run_context.go [hunk 2]
```diff
@@ -167,6 +215,31 @@ func (rc *RunContext) stopJobContainer() common.Executor {
...
 
+func (rc *RunContext) startServiceContainers() common.Executor {
+	return func(ctx context.Context) error {
+		execs := []common.Executor{}
+		for _, c := range rc.ServiceContainers {
+			execs = append(execs, common.NewPipelineExecutor(
+				c.Pull(false),
+				c.Create(),
+				c.Start(false),
+				c.ConnectToNetwork(defaultNetwork),
```
Comment (catthehacker) [inline pkg/runner/run_context.go:226]: Network name clash.

File: pkg/runner/run_context.go [hunk 3]
```diff
@@ -130,9 +158,14 @@ func (rc *RunContext) startJobContainer() common.Executor {
...
 		return common.NewPipelineExecutor(
 			rc.JobContainer.Pull(rc.Config.ForcePull),
+			rc.stopServiceContainers(),
 			rc.stopJobContainer(),
+			rc.removeNetwork(),
+			rc.createNetwork(),
+			rc.startServiceContainers(),
 			rc.JobContainer.Create(),
 			rc.JobContainer.Start(false),
+			rc.JobContainer.ConnectToNetwork(defaultNetwork),
```
Comment (catthehacker) [inline pkg/runner/run_context.go:168]: Network name clash.

**Review by catthehacker — CHANGES_REQUESTED** [review]
Anything else looks good to me!

File: pkg/container/docker_network.go [hunk 4]
```diff
@@ -14,10 +15,11 @@ func NewDockerNetworkCreateExecutor(name string) common.Executor {
 			return err
 		}
 
-		_, err = cli.NetworkCreate(ctx, name, types.NetworkCreate{})
+		network, err := cli.NetworkCreate(ctx, name, types.NetworkCreate{})
 		if err != nil {
 			return err
 		}
+		fmt.Printf("%#v", network.ID)
```
Comment (catthehacker) [inline pkg/container/docker_network.go:22]: Please use `log` instead of `fmt`, we should limit `fmt` only to formatting strings

**Review by catthehacker — CHANGES_REQUESTED** [review]

File: pkg/container/docker_network.go (hunk 4 above)
Comment (Markcial) [reply pkg/container/docker_network.go:22]: Oh sorry, it was a rogue line, my apologies

File: pkg/container/docker_network.go [hunk 5]
```diff
@@ -2,6 +2,7 @@ package container
 
 import (
 	"context"
+	"fmt"
```
Comment (Markcial) [inline pkg/container/docker_network.go:5]:
```suggestion
```

File: pkg/container/docker_network.go (hunk 4 above)
Comment (Markcial) [inline pkg/container/docker_network.go:22]:
```suggestion
```

File: pkg/runner/run_context.go [hunk 6]
```diff
@@ -96,6 +97,36 @@ func (rc *RunContext) startJobContainer() common.Executor {
 			binds = append(binds, fmt.Sprintf("%s:%s%s", rc.Config.Workdir, rc.Config.Workdir, bindModifiers))
 		}
 
+		if rc.Config.ContainerArchitecture == "" {
+			rc.Config.ContainerArchitecture = fmt.Sprintf("%s/%s", "linux", runtime.GOARCH)
+		}
```
Comment (catthehacker) [inline pkg/runner/run_context.go:102]: This seems like old code re-introduced through merge

**Review by catthehacker — CHANGES_REQUESTED** [review]

File: pkg/runner/run_context.go (hunk 6 above)
Comment (Markcial) [reply pkg/runner/run_context.go:102]:
```suggestion
```

File: pkg/runner/runner_test.go [hunk 7]
```diff
@@ -153,6 +153,42 @@ func TestRunEventSecrets(t *testing.T) {
...
+	runnerConfig := &Config{
+		Workdir:         workdir,
+		EventName:       eventName,
+		Platforms:       platforms,
+		ReuseContainers: false,
+	}
+	runner, err := New(runnerConfig)
+	assert.NilError(t, err, workflowPath)
+
+	planner, err := model.NewWorkflowPlanner(fmt.Sprintf("testdata/%s", workflowPath))
```
Comment (catthehacker) [inline pkg/runner/runner_test.go:183]: `NewWorkflowPlanner` was changed recently, below is something you probably want
```suggestion
	planner, err := model.NewWorkflowPlanner(fmt.Sprintf("testdata/%s", workflowPath, true))
```

**Review by catthehacker — CHANGES_REQUESTED** [review]

File: pkg/runner/runner_test.go [hunk 8]
```diff
@@ -153,6 +153,42 @@ func TestRunEventSecrets(t *testing.T) {
...
+	runnerConfig := &Config{
+		Workdir:         workdir,
+		EventName:       eventName,
+		Platforms:       platforms,
+		ReuseContainers: false,
+	}
+	runner, err := New(runnerConfig)
+	assert.NilError(t, err, workflowPath)
+
+	planner, err := model.NewWorkflowPlanner(fmt.Sprintf("testdata/%s", workflowPath, true))
```
Comment (catthehacker) [inline pkg/runner/runner_test.go:183]: Actually my bad, should be
```suggestion
	planner, err := model.NewWorkflowPlanner(fmt.Sprintf("testdata/%s", workflowPath), true)
```

**Review by catthehacker — CHANGES_REQUESTED** [review]

**Comment by catthehacker** [comment]
It seems that it created 5 different `act-test-network` networks
```
[basic/build]   🐳  docker pull node:12.20.1-buster-slim
time="2021-05-04T09:27:22Z" level=debug msg="Image exists? true"
[basic/build] Removed container: 6648844bde9265313d5fed13c0dbeaded7745c5deb5db238bb8ee514bf595278
[basic/build]   🐳  docker volume rm act-basic-build
[basic/build]   🐳  docker create image=node:12.20.1-buster-slim platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[basic/build] Created container name=act-basic-build id=d2af12ba6bb5dee3e5133bf1c8690d62ffc99b46397813da006a6b0d056e5570 from image node:12.20.1-buster-slim (platform: )
[basic/build] ENV ==> [RUNNER_TOOL_CACHE=/opt/hostedtoolcache RUNNER_OS=Linux RUNNER_TEMP=/tmp]
[basic/build]   🐳  docker run image=node:12.20.1-buster-slim platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[basic/build] Starting container: d2af12ba6bb5dee3e5133bf1c8690d62ffc99b46397813da006a6b0d056e5570
[basic/build] Started container: d2af12ba6bb5dee3e5133bf1c8690d62ffc99b46397813da006a6b0d056e5570
[basic/build]   🐳  docker network connect act-build-network act-basic-build
time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
--- FAIL: TestRunEventPullRequest (0.80s)
    runner_test.go:226: assertion failed: error is not nil: Error response from daemon: network act-build-network is ambiguous (5 matches found on name): pull-request
```

**Comment by Markcial** [comment]
> It seems that it created 5 different `act-test-network` networks
> 
> ```
> [basic/build]   🐳  docker pull node:12.20.1-buster-slim
> time="2021-05-04T09:27:22Z" level=debug msg="Image exists? true"
> [basic/build] Removed container: 6648844bde9265313d5fed13c0dbeaded7745c5deb5db238bb8ee514bf595278
> [basic/build]   🐳  docker volume rm act-basic-build
> [basic/build]   🐳  docker create image=node:12.20.1-buster-slim platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
> [basic/build] Created container name=act-basic-build id=d2af12ba6bb5dee3e5133bf1c8690d62ffc99b46397813da006a6b0d056e5570 from image node:12.20.1-buster-slim (platform: )
> [basic/build] ENV ==> [RUNNER_TOOL_CACHE=/opt/hostedtoolcache RUNNER_OS=Linux RUNNER_TEMP=/tmp]
> [basic/build]   🐳  docker run image=node:12.20.1-buster-slim platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
> [basic/build] Starting container: d2af12ba6bb5dee3e5133bf1c8690d62ffc99b46397813da006a6b0d056e5570
> [basic/build] Started container: d2af12ba6bb5dee3e5133bf1c8690d62ffc99b46397813da006a6b0d056e5570
> [basic/build]   🐳  docker network connect act-build-network act-basic-build
> time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
> time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
> time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
> time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
> time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
> time="2021-05-04T09:27:23Z" level=debug msg="Error response from daemon: network act-build-network is ambiguous (5 matches found on name)"
> --- FAIL: TestRunEventPullRequest (0.80s)
>     runner_test.go:226: assertion failed: error is not nil: Error response from daemon: network act-build-network is ambiguous (5 matches found on name): pull-request
> ```

Yes, this was one of my concerns, I was investigating on how to have the network ID, because do the operations just with the network name is a bit flaky.

I have a busy week, but if you can wait until next monday I can dig on this

**Comment by catthehacker** [comment]
I kind of resolved it but then stuck with issue of too much IP ranges allocated

**Comment by catthehacker** [comment]
thank you bot, I know that -.-

**Comment by Markcial** [comment]
If you get something let me know, as I can't get to it before next monday.

I have an idea about using the network identifier to keep track on the networks and how to specifically remove them. But until I get some spare time I won't be able to try it.

Thanks for getting some time into it

**Comment by catthehacker** [comment]
https://github.com/catthehacker/act-fork/tree/ticket-173
There is a lot in there unrelated changes as well (since I've a bad habit of going on refactoring spree when looking through code)

**Comment by Markcial** [comment]
> https://github.com/catthehacker/act-fork/tree/ticket-173
> There is a lot in there unrelated changes as well (since I've a bad habit of going on refactoring spree when looking through code)

Well don't worry, at least it's something to check for ideas, thanks

**Pull request closed without merging**